### PR TITLE
stats: allow combining configured tags with custom tags

### LINKED ISSUE #27030: stats: allow combining configured tags with those supplied to counterFromStatNameWithTags
The problem I'm facing is being able to add custom tags to a certain metrics, which their value is dynamic. Taking a counter stat for example, I would like to add a tag which its value is only being known at runtime. Ideally, when calling ``inc()``, on a counter, I would also like to provide that tag: ``inc(key, value)`` or ``inc(set_of_tags_with_values)``.

An example use case (although I wouldn't do this, but just to show what I mean in practice), is a call for ``inc("time", getCurrentTime())``, which is placed in ``onNewConnection()``. Here the time value varies, so every time ``onNewConnection()`` is being called, it gets a different value

Would it be possible to add an overload to ``inc()`` as in the example above to cope with this? Otherwise, I'll appreciate help figuring out this problem.

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message: stats: allow combining configured tags with custom tags
Additional Description: Adds a way to create a StatNameTagVector that contains extracted tags names from both the configured and well defined tags, and custom provided tags. Resolves #27030.
Risk Level: Low
Testing: Unit tests
Docs Changes: None
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

**Review by jmarantz — COMMENTED** [review]
At a high level this looks good! I have a few nits I can add later. But can we add some unit-tests to see the whole flow?

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:778]: can you also make a test where you are using the regex-extracted tags?

To do this you'll see to calll
```
  store_->setTagProducer(std::make_unique<Stats::TagProducerImpl>(stats_config));
```
if that's not already happened. and then make a stat like "cluster.cluster_name.leaf"

**Review by jmarantz — COMMENTED** [review]
looks looks great!

A few minor comments but basically this should be close to ready.

File: test/common/stats/thread_local_store_test.cc
Comment (ohadvano) [reply test/common/stats/thread_local_store_test.cc:778]: The two tests below this one are regex testing, unless I misunderstood what you mean?

File: source/common/stats/thread_local_store.cc
Comment (ohadvano) [reply source/common/stats/thread_local_store.cc]: done, thanks that's useful.

File: test/integration/server.h
Comment (ohadvano) [reply test/integration/server.h]: done

File: envoy/stats/store.h
Comment (ohadvano) [reply envoy/stats/store.h]: done

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [reply test/common/stats/thread_local_store_test.cc:778]: Sorry, I meant using the builtin expressions (not necessarily regexes actually) to automatically extract things like cluster.foo.bar into tag-extracted name "bar" with tag name="cluster" value="bar", based on the token and regex patterns around https://github.com/envoyproxy/envoy/blob/f88e29647931c7c1f1d2ff181e4d95602b0928fa/source/common/config/well_known_names.cc#L128

Comment (ohadvano) [reply test/common/stats/thread_local_store_test.cc:778]: Got it, added.

File: test/common/stats/thread_local_store_test.cc
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:865]: nit: this should be ASSERT_EQ -- that's the pattern we use for a check that, if it fails, the test-method should exit right after that to avoid crashing and thus not running the rest of the tests.

The others can be EXPECT_EQ.

File: source/common/stats/isolated_store_impl.h
Comment (jmarantz) [inline source/common/stats/isolated_store_impl.h:238]: s/not in use/not yet implemented/g

**Review by jmarantz — COMMENTED** [review]
lgtm modulo some minor comments.

/assign-from @envoyproxy/senior-maintainers

File: source/common/stats/isolated_store_impl.h
Comment (ohadvano) [reply source/common/stats/isolated_store_impl.h]: done

File: test/common/stats/thread_local_store_test.cc
Comment (ohadvano) [reply test/common/stats/thread_local_store_test.cc]: done

**Review by jmarantz — APPROVED** [review]

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
